coderd/workspaceapps/token.go

```diff
@@ -0,0 +1,206 @@
+package workspaceapps
+
+import (
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/json"
+	"time"
+
+	"github.com/google/uuid"
+	"golang.org/x/xerrors"
+	"gopkg.in/square/go-jose.v2"
+
+	"github.com/coder/coder/coderd/database"
+)
+
+const tokenSigningAlgorithm = jose.HS512
+
+// SignedToken is the struct data contained inside a workspace app JWE. It
+// contains the details of the workspace app that the token is valid for to
+// avoid database queries.
+type SignedToken struct {
+	// Request details.
+	Request `json:"request"`
+
+	// Trusted resolved details.
+	Expiry      time.Time `json:"expiry"` // set by GenerateToken if unset
+	UserID      uuid.UUID `json:"user_id"`
+	WorkspaceID uuid.UUID `json:"workspace_id"`
+	AgentID     uuid.UUID `json:"agent_id"`
+	AppURL      string    `json:"app_url"`
+}
+
+// MatchesRequest returns true if the token matches the request. Any token that
+// does not match the request should be considered invalid.
+func (t SignedToken) MatchesRequest(req Request) bool {
+	return t.AccessMethod == req.AccessMethod &&
+		t.BasePath == req.BasePath &&
+		t.UsernameOrID == req.UsernameOrID &&
+		t.WorkspaceNameOrID == req.WorkspaceNameOrID &&
+		t.AgentNameOrID == req.AgentNameOrID &&
+		t.AppSlugOrPort == req.AppSlugOrPort
+}
+
+// AppSigningKey is used for signing and encrypting app tokens and API keys.
+// TODO: we cannot use the same key for signing and encrypting with two
+// different algorithms, that's a security issue. We should use a different key
+// for each.
+// OR
+// We get rid of signing and use encryption for both api keys and tickets.
+// Do this by switching to JWE.
+type AppSigningKey [64]byte
```

Comment:
This should be called `SigningKey` instead because it's already in `workspaceapps`